cni/pkg/taint/taint.go

```diff
@@ -0,0 +1,154 @@
+	ConfigmapNamespace string
+}
+type TaintSetter struct {
+	configs []ConfigSettings // contains all configmaps information
+	Client  client.Interface
+}
+
+func (ts TaintSetter) GetAllConfigs() []ConfigSettings {
+	return ts.configs
+}
+func NewTaintSetter(clientset client.Interface, options *Options) (ts TaintSetter, err error) {
+	configmap, err := clientset.CoreV1().ConfigMaps(options.ConfigmapNamespace).Get(context.TODO(), options.ConfigmapName, metav1.GetOptions{})
+	if err != nil {
+		return ts, err
+	}
+	ts = TaintSetter{
+		configs: []ConfigSettings{},
+		Client:  clientset,
+	}
+	ts.LoadConfig(*configmap)
+	return
+}
+
+// load corresponding configmap's critical labels and their namespace
+func (ts *TaintSetter) LoadConfig(config v1.ConfigMap) {
+	log.Debugf("Loading configmap %s in %s", config.Name, config.Namespace)
+	ts.configs = make([]ConfigSettings, 0) //clear previous one
+	for key, value := range config.Data {
+		log.Debugf("Loading %s ", key)
+		var tempset ConfigSettings
+		err := yaml.Unmarshal([]byte(value), &tempset)
+		if err != nil {
+			log.Fatalf("cannot unmarshal data : %v", err)
+		}
+		_, err = metav1.ParseToLabelSelector(tempset.LabelSelector)
+		if err != nil {
+			log.Fatalf("not a valid label selector: %v", err)
+		}
+		ts.configs = append(ts.configs, tempset)
+		log.Infof("successfully loaded %s", tempset.ToString())
+	}
+}
+
+//only pod with NodeReadiness Toleracnce with effect no schedule or
+//a generalized tolerance with noschedule effect can be considered
+func (ts TaintSetter) validTolerance(pod v1.Pod) bool {
+	for _, toleration := range pod.Spec.Tolerations {
+		if (toleration.Key == TaintName || toleration.Key == "") &&
+			toleration.Operator == v1.TolerationOpExists &&
+			toleration.Effect == v1.TaintEffectNoSchedule {
+			return true
+		}
+	}
+	return false
+}
+
+//check whether current node have readiness
+func (ts TaintSetter) HasReadinessTaint(node *v1.Node) bool {
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName && taint.Effect == v1.TaintEffectNoSchedule {
+			return true
+		}
+	}
+	return false
+}
+
+//assumption: order of taint is not important
+func (ts TaintSetter) RemoveReadinessTaint(node *v1.Node) error {
+	updatedTaint := deleteTaint(node.Spec.Taints, &v1.Taint{Key: TaintName, Effect: v1.TaintEffectNoSchedule})
+	node.Spec.Taints = updatedTaint
+	updatedNodeWithTaint, err := ts.Client.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil || updatedNodeWithTaint == nil {
+		return fmt.Errorf("failed to update node %v after adding taint: %v", node.Name, err)
+	}
+	log.Infof("Successfully removed taint on node %v", updatedNodeWithTaint.Name)
+	return nil
+}
+
+// taint node with specific taint name with effect of no schedule
+//do nothing if it already have the readiness taint
+func (ts TaintSetter) AddReadinessTaint(node *v1.Node) error {
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == TaintName {
```

Comment:
what if effect is not equal to `v1.TaintEffectNoSchedule`? is that possible?

Reply:
added it to logic